src/Disks/ObjectStorages/Encrypted/registerDiskEncryptedOS.cpp

```diff
@@ -0,0 +1,135 @@
+#    include "EncryptedObjectStorage.h"
+#    include <IO/FileEncryptionCommon.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+}
+
+static EncryptedObjectStorageSettingsPtr parseDiskEncryptedOSSettings(
+    const String & disk_name, const Poco::Util::AbstractConfiguration & config, const String & config_prefix, const DisksMap & disk_map)
+{
+    try
+    {
+        auto ret = std::make_shared<EncryptedObjectStorageSettings>();
+
+        std::map<UInt64, String> keys_by_id;
+        Strings keys_without_id;
+        FileEncryption::getKeysFromConfig(config, config_prefix, keys_by_id, keys_without_id);
+
+        for (const auto & [key_id, key] : keys_by_id)
+        {
+            auto fingerprint = FileEncryption::calculateKeyFingerprint(key);
+            ret->all_keys[fingerprint] = key;
+
+            /// Version 1 used key fingerprints based on the key id.
+            /// We have to add such fingerprints to the map too to support reading files encrypted by version 1.
+            auto v1_fingerprint = FileEncryption::calculateV1KeyFingerprint(key, key_id);
+            ret->all_keys[v1_fingerprint] = key;
+        }
+
+        for (const auto & key : keys_without_id)
+        {
+            auto fingerprint = FileEncryption::calculateKeyFingerprint(key);
+            ret->all_keys[fingerprint] = key;
+        }
+
+        String current_key = FileEncryption::getCurrentKeyFromConfig(config, config_prefix, keys_by_id, keys_without_id);
+        ret->current_key = current_key;
+        ret->current_key_fingerprint = FileEncryption::calculateKeyFingerprint(current_key);
+        ret->current_algorithm = FileEncryption::getCurrentAlgorithmFromConfig(config, config_prefix);
+
+        FileEncryption::checkKeySize(ret->current_key.size(), ret->current_algorithm);
+
+        auto wrapped_disk_name = config.getString(config_prefix + ".disk", "");
+        if (wrapped_disk_name.empty())
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Disk EncryptedOS requires `disk` field in config");
+
+        auto disk_it = disk_map.find(wrapped_disk_name);
+        if (disk_it == disk_map.end())
+        {
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Cannot wrap disk `{}` with encryption layer `{}`: there is no such disk (it should be initialized before encryption disk)",
+                wrapped_disk_name,
+                disk_name);
+        }
+        ret->wrapped_disk = disk_it->second;
+        if (!dynamic_cast<const DiskObjectStorage *>(ret->wrapped_disk.get()))
+            throw Exception(
+                ErrorCodes::BAD_ARGUMENTS,
+                "Cannot wrap disk `{}` with encryption layer `{}`: encrypted_os disk is allowed only on top of object storage",
+                wrapped_disk_name,
+                disk_name);
+
+        auto header_cache_path = config.getString(config_prefix + ".header_cache_path", "");
+        if (!header_cache_path.empty())
```

Comment:
why do we need a separate cache for header?

Reply:
May be reuse the same cache object which is used for caching the data? It is already accessible where you use it.

Reply:
Looks like it is fixable by passing cache from CachedObjectStorage via write settings settings similar how it is done on read. But yes, it is kind of ugly and makes this abstraction of IObjectStorage leak again :(. Therefore the way you did it initially - with a separate cache object - makes sense. Let me think. Will respond a bit later today.

Reply:
@aiven-sal, let's make it as it is now, but instead of specifying a full cache configuration for the header cache and making a new cache object - let's make a single config parameter `<header_cache>cache_name</header_cache>`, where `cache_name` defines which of the existing caches to use (you can get a cache object by cache name from `FileCacheFactory::instance()`).

Reply:
As I understand all you need is patching ReadBuffer and WriteBuffer for header reading/writing, so all you need is creating CachedOnDiskRead(Write)BufferFromFile over another buffer and knowing which cache object you need to use for them, which only requires the cache name, then you get the cache object from the cache factory. E.g. basically it will be the same as it is now, but with the difference that the cache object is not created specifically for header but already exists. About what changes in configuration, not much, just one more configuration parameter for encrypted disk - cache_name, which is always checked, but noop in case we do not create encrypted object storage.